20453 show sources sinks in cluster

calumrussell opened the pull request:
Adds support for IN CLUSTER condition to SHOW SOURCES and SHOW SINKS.

### Motivation

Closes (hopefully) https://github.com/MaterializeInc/materialize/issues/20453

### Tips for reviewer

My first PR on Materialize so I am not sure if I have missed something.

Majority of changes are in src/sql-parser and have tested changes/output using that package.

Also changed the doc/user for both statements.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  Add the optional `IN CLUSTER` filter to `SHOW SOURCES` and `SHOW SINKS` that filters on cluster.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=21015) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=21015) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=21015) it.</sub>

calumrussell commented:
I was unable to complete this PR fully because the tests were too resource-intensive for my computer.

I believe the core logic is complete here but:
- the SHOW SOURCES query in misc/python/materialize/checks/identifiers.py needs the completing change (this is small but I was unable to test this locally at all so wasn't able to finish it).
- an outer join may be needed in `show_sources` in sql/src/plan/statement/show.rs to fix some of the cases in testdrive that run queries against materialize system tables
- sql output in testdrive still needs to be changed as some of the clusters used weren't default (but, again, I was unable to test this fully locally).

jkosh44 commented on misc/python/materialize/checks/identifiers.py:
```diff
@@ -272,9 +272,9 @@ def validate(self) -> Testdrive:
-        {dq_print(self.ident["sink0"])} kafka ${{arg.default-storage-size}}
-        {dq_print(self.ident["sink1"])} kafka ${{arg.default-storage-size}}
-        {dq_print(self.ident["sink2"])} kafka ${{arg.default-storage-size}}
+        {dq_print(self.ident["sink0"])} kafka ${{arg.default-storage-size}} {dq(self.ident["db"])}_{dq(self.ident["schema"])}_{dq(self.ident["sink0"])}
+        {dq_print(self.ident["sink1"])} kafka ${{arg.default-storage-size}} {dq(self.ident["db"])}_{dq(self.ident["schema"])}_{dq(self.ident["sink1"])}
+        {dq_print(self.ident["sink2"])} kafka ${{arg.default-storage-size}} {dq(self.ident["db"])}_{dq(self.ident["schema"])}_{dq(self.ident["sink2"])}
```
For some reason it looks like in the test output there's spaces inbetween the db, schema, and idents:
```
Poor diff:
- 15.0 kafka 4-1 -1.0 _0.0 _15.0
+ 15.0 kafka 4-1 -1.0_0.0_15.0
- 16.0 kafka 4-1 -1.0 _0.0 _16.0
+ 16.0 kafka 4-1 -1.0_0.0_16.0
- 17.0 kafka 4-1 -1.0 _0.0 _17.0
+ 17.0 kafka 4-1 -1.0_0.0_17.0
```

jkosh44 reviewed (commented):
Thanks for the contribution! 

The SQL Logic Test output is very noisy, but if you download the output and search for "fail:" you should find the failures and they're usually straight forward to fix.

Similarly if you search for "error:" in the Testdrive output the failures are usually pretty straight forward to fix.

Unfortunately I won't be able to do a deep dive on each test failure for now.

calumrussell commented:
Remaining issue with testdrive/postgres CDC/sqllogictest appear to all be linked to rows being removed by the SHOW SOURCES command.

From testdrive/load-generator.td test output:

`
> SHOW SOURCES
--
  | rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 22s
  | load-generator.td:12:1: error: non-matching rows: expected:
  | [["accounts", "subsource", "<null>", "default"], ["auction_house", "load-generator", "4-1", "default"], ["auction_house_progress", "progress", "<null>", "default"], ["auctions", "subsource", "<null>", "default"], ["bids", "subsource", "<null>", "default"], ["organizations", "subsource", "<null>", "default"], ["users", "subsource", "<null>", "default"]]
  | got:
  | [["auction_house", "load-generator", "4-1", "materialize_public_auction_house"]]
  | Poor diff:
  | - accounts subsource <null> default
  | - auction_house load-generator 4-1 default
  | + auction_house load-generator 4-1 materialize_public_auction_house
  | - auction_house_progress progress <null> default
  | - auctions subsource <null> default
  | - bids subsource <null> default
  | - organizations subsource <null> default
  | - users subsource <null> default
`

From load-generator.td:
`
> CREATE SOURCE auction_house FROM LOAD GENERATOR AUCTION FOR ALL TABLES;

> SHOW SOURCES
accounts                subsource      <null>                         default
auction_house           load-generator ${arg.default-storage-size}    default
auction_house_progress  progress       <null>                         default
auctions                subsource      <null>                         default
bids                    subsource      <null>                         default
organizations           subsource      <null>                         default
users                   subsource      <null>                         default`
`

Query used for SHOW SOURCES:
`
let query = format!(
        "SELECT sources.name, sources.type, sources.size, clusters.name as cluster
        FROM mz_catalog.mz_sources AS sources
        JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sources.cluster_id
        WHERE {where_clause}"
    );
`

All of the tests appear to be showing rows being removed, so my assumption was that the join was removing rows but making this change caused testdrive to fail everywhere with:
`
error: deadline has elapsed
`
I don't see what it can be other than the JOIN statement removing rows but I am not totally familiar with the structure of mz_catalog tables either, so may be missing something else. `loadgenerator` source types seem fine but `progress` and `subsource` were being removed.

jkosh44 commented:
> Query used for SHOW SOURCES by `show_sources` in plan/statement/show.rs:
> 
> ```
> let query = format!(
>         "SELECT sources.name, sources.type, sources.size, clusters.name as cluster
>         FROM mz_catalog.mz_sources AS sources
>         JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sources.cluster_id
>         WHERE {where_clause}"
>     );
> ```

I have not looked into the test failures in depth yet, but have you tried `LEFT JOIN`? As in

```
 let query = format!(
         "SELECT sources.name, sources.type, sources.size, clusters.name as cluster
         FROM mz_catalog.mz_sources AS sources
         LEFT JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sources.cluster_id
         WHERE {where_clause}"
     );
```

calumrussell commented:
@jkosh44 I have been unable to work out where the problem with identifiers.py is coming from. It seems to be something in the test env because I have tested the file by itself and couldn't reproduce the output.

jkosh44 commented:
> @jkosh44 I have been unable to work out where the problem with identifiers.py is coming from. It seems to be something in the test env because I have tested the file by itself and couldn't reproduce the output.

Sorry for the delay @calumrussell I was on vacation. I'll take a look and see if I can resolve the issue in identifiers.py.

calumrussell commented:
> > @jkosh44 I have been unable to work out where the problem with identifiers.py is coming from. It seems to be something in the test env because I have tested the file by itself and couldn't reproduce the output.
> 
> Sorry for the delay @calumrussell I was on vacation. I'll take a look and see if I can resolve the issue in identifiers.py.

I checked Python's dedent and tried a few variations with the formatting of the fstring. Half of the issue appeared to be solved by moving the lookups inside {} (which I think was related to fstring and the dq function, which returned an fstring itself). But then a space began appearing on the end.

I was unable to work out how the code was being called so wasn't sure if the space was being added somewhere else. I checked dedent a couple of times, which is the only external call within that file, so my assumption was that the space is being added somewhere else.

jkosh44 commented:
@calumrussell OK, it looks like this is going to pass CI now, or at least get past the previous errors. The `unknown connection: Dynamic(IdHandleInner { allocator: IdAllocator(Mutex { data: IdAllocatorInner { next: 3, max: 65536, free: [] }, poisoned: false, .. }), id: 2 })` error is a known flake. Will you mark this as ready for review and I'll give this a more in depth review?

One thing I noticed while fixing the test is that we need to update the builtin indexes. We automatically create indexes in Materialize to help serve the `SHOW` commands, for example:
https://github.com/MaterializeInc/materialize/blob/f63c99bb12b16bfa4b6a2a3d95a4328fa0c9bc68/src/adapter/src/catalog/builtin.rs#L4643-L4650
https://github.com/MaterializeInc/materialize/blob/f63c99bb12b16bfa4b6a2a3d95a4328fa0c9bc68/src/adapter/src/catalog/builtin.rs#L4670-L4677

The show sinks and show sources indexes should be updated to include the cluster ID (similar to how the show materialized views index works).

jkosh44 reviewed (changes requested):
> One thing I noticed while fixing the test is that we need to update the builtin indexes.

I actually misspoke here. What we actually want to do is create views for the joins performed in `SHOW SINKS` and `SHOW SOURCES`. You can use Materialized Views as a reference:

https://github.com/MaterializeInc/materialize/blob/f63c99bb12b16bfa4b6a2a3d95a4328fa0c9bc68/src/adapter/src/catalog/builtin.rs#L4218-L4225

Then the indexes you create should be created on the views, not the underlying tables.

Everything else LGTM.

calumrussell commented:
> > One thing I noticed while fixing the test is that we need to update the builtin indexes.
> 
> I actually misspoke here. What we actually want to do is create views for the joins performed in `SHOW SINKS` and `SHOW SOURCES`. You can use Materialized Views as a reference:
> 
> https://github.com/MaterializeInc/materialize/blob/f63c99bb12b16bfa4b6a2a3d95a4328fa0c9bc68/src/adapter/src/catalog/builtin.rs#L4218-L4225
> 
> Then the indexes you create should be created on the views, not the underlying tables.
> 
> Everything else LGTM.

Ah, should the code in plan/statement/show.rs also be querying the view? Taking SHOW SINKS as an example, should this section query the view created in catalog/builtin.rs:

https://github.com/MaterializeInc/materialize/blob/cf83758b739176f23c7cf648b524e93a8ac30c26/src/sql/src/plan/statement/show.rs#L484-L516

And then use those views for the indexes too.

jkosh44 commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -4215,6 +4215,24 @@ WHERE false
+    sql: "CREATE VIEW mz_internal.mz_show_sources
+AS SELECT sources.name, sources.type, sources.size, clusters.name as cluster, schema_id, cluster_id,
+FROM mz_catalog.mz_sources AS sources
+LEFT JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sources.cluster_id",
```
Can you remind me, why are these `LEFT JOIN`s instead of just `JOIN`s?

calumrussell replied:
I believe the issue I had with the JOIN on mz_clusters was that I was using OUTER JOIN, so changing to JOIN should be fine.

I thought about changing this but I can't run sqllogictest locally atm so didn't want to make more changes I couldn't test.

jkosh44 replied:
I think we should change this to `JOIN` then to match the materialized view case. I'll make sure to kick off the CI build which should hopefully let us know if something failed.

jkosh44 commented on src/adapter/src/catalog/builtin.rs:
These need to get added to the vec `BUILTINS_STATIC` at the bottom of the file.

jkosh44 commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -4215,6 +4215,24 @@ WHERE false
     ",
 };
 
+pub const MZ_SHOW_SOURCES: BuiltinView = BuiltinView {
+    name: "mz_show_sources",
+    schema: MZ_INTERNAL_SCHEMA,
+    sql: "CREATE VIEW mz_internal.mz_show_sources
+AS SELECT sources.name, sources.type, sources.size, clusters.name as cluster, schema_id, cluster_id,
+FROM mz_catalog.mz_sources AS sources
+JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sources.cluster_id",
+};
+
+pub const MZ_SHOW_SINKS: BuiltinView = BuiltinView {
+    name: "mz_show_sinks",
+    schema: MZ_INTERNAL_SCHEMA,
+    sql:"CREATE VIEW mz_internal.mz_show_sinks
+AS SELECT sinks.name, sinks.type, sinks.size, clusters.name as cluster, schema_id, cluster_id,
+FROM mz_catalog.mz_sinks AS sinks
+JOIN mz_catalog.mz_clusters AS clusters ON clusters.id = sinks.cluster_id",
+};
+
```
These need to get added to the vec `BUILTINS_STATIC` at the bottom of the file.

jkosh44 commented:
@calumrussell I'm going to try and resolve the new test failures. They all seem pretty straightforward, but will be annoying for you to resolve without being able to run them locally.

calumrussell commented:
> @calumrussell I'm going to try and resolve the new test failures. They all seem pretty straightforward, but will be annoying for you to resolve without being able to run them locally.

I was able to run them. Am just waiting on one fail with testdrive.

I should be able to push in an hour or so.

calumrussell commented:
The LEFT JOIN seems to make a difference for SHOW SOURCES, I am not sure why but I noticed that other sources types (for example, subsources) were missing from results of SHOW SOURCES if a plain JOIN was used.

I am fairly sure that slt should be fine here but I also had some odd results from testdrive that appeared to be unrelated to changes made.

jkosh44 commented:
> The LEFT JOIN seems to make a difference for SHOW SOURCES, I am not sure why but I noticed that other sources types (for example, subsources) were missing from results of SHOW SOURCES if a plain JOIN was used.
> 
> I am fairly sure that slt should be fine here but I also had some odd results from testdrive that appeared to be unrelated to changes made.

Ah yeah, subsources and progress sources will not have a cluster ID. So LEFT JOIN is the right way to go.

jkosh44 commented on test/sqllogictest/information_schema_tables.slt:
```diff
@@ -643,6 +643,15 @@ materialize
+mz_internal
+mz_show_sources
+VIEW
+
```
You have an extra newline here.

calumrussell commented:
Catalog.td in sqllogictest has:

```
_bk;t=16935892403842023-09-01T17:27:20.383814Z  INFO tokio_postgres::connection: NOTICE: relation "consensus" already exists, skipping    

_bk;t=1693589240681    EXPLAIN SHOW SOURCES

_bk;t=1693589240681    OutputFailure:test/sqllogictest/system-cluster.slt:68

_bk;t=1693589240681            expected: Values(["Explained Query (fast path):\n  Project (#2..=#5)\n    Filter (#0 = \"u3\")\n      ReadExistingIndex mz_internal.mz_show_sources_ind\n\nUsed Indexes:\n  - mz_internal.mz_show_sources_ind (*** full scan ***)\n"])

_bk;t=1693589240681            actually: Values(["Explained Query (fast path):\n  Project (#2..=#5)\n    Filter (#0 = \"u3\")\n      ReadExistingIndex mz_internal.mz_show_sources_ind\n\nUsed Indexes:\n  - mz_internal.mz_show_sources_ind (*** full scan ***)\n\nNotices:\n  - Notice: Index mz_internal.mz_show_sources_ind on mz_show_sources(schema_id, cluster_id) is too wide to use for literal equalities `schema_id = \"u3\"`.\n    Hint: If your literal equalities filter out many rows, create an index whose key exactly matches your literal equalities: (schema_id).\n"])

_bk;t=1693589240681            actual raw: [Row { columns: [Column { name: "Optimized Plan", type: Text }] }]

_bk;t=1693589240681    ----

_bk;t=1693589240700    EXPLAIN SHOW SINKS

_bk;t=1693589240700    OutputFailure:test/sqllogictest/system-cluster.slt:144

_bk;t=1693589240700            expected: Values(["Explained Query (fast path):\n  Project (#2..=#5)\n    Filter (#0 = \"u3\")\n      ReadExistingIndex mz_internal.mz_show_sinks_ind\n\nUsed Indexes:\n  - mz_internal.mz_show_sinks_ind (*** full scan ***)\n"])

_bk;t=1693589240700            actually: Values(["Explained Query (fast path):\n  Project (#2..=#5)\n    Filter (#0 = \"u3\")\n      ReadExistingIndex mz_internal.mz_show_sinks_ind\n\nUsed Indexes:\n  - mz_internal.mz_show_sinks_ind (*** full scan ***)\n\nNotices:\n  - Notice: Index mz_internal.mz_show_sinks_ind on mz_show_sinks(schema_id, cluster_id) is too wide to use for literal equalities `schema_id = \"u3\"`.\n    Hint: If your literal equalities filter out many rows, create an index whose key exactly matches your literal equalities: (schema_id).\n"])

_bk;t=1693589240700            actual raw: [Row { columns: [Column { name: "Optimized Plan", type: Text }] }]

_bk;t=1693589240700    ----
```

The error suggests this may relate to adding cluster_id to the indexes for SHOW SINKS/SHOW SOURCES but I didn't see more than one param to the Filter on SHOW MATERIALIZED VIEWS?

jkosh44 commented:
> The error suggests this may relate to adding cluster_id to the indexes for SHOW SINKS/SHOW SOURCES but I didn't see more than one param to the Filter on SHOW MATERIALIZED VIEWS in the same slt file?

Ah, looking at the show materialized views index, it looks like we exclude cluster id from the index. We should probably do the same for sinks and sources.

calumrussell commented:
I had time to go through the fails for testdrive and it is returning a different error every time.

For example with catalog.td line 663 initially was:
```
+ 147
- 99
```
then when I changed it and nothing else:
```
+ 123
- 147
```
Same thing is happening with a few other tests. I initially ran the test and got (I think) two good fails, and three spurious ones. I fixed, ran tests again, and then 11 files failed.

Slt was passing, and the ones I completed for testdrive are the ones that seemed to be correct.

jkosh44 commented:
How did you run the Testdrive tests? The testdrive tests are somewhat sensitive to how they're run, so if it's not done properly there can be some false negatives.

The non-nightlies are passing so I think it's fine. The nightlies are still failing due to some issue with escaping characters.

calumrussell commented:
> How did you run the Testdrive tests? The testdrive tests are somewhat sensitive to how they're run, so if it's not done properly there can be some false negatives.
> 
> The non-nightlies are passing so I think it's fine. The nightlies are still failing due to some issue with escaping characters.

Ah, I was running ./mzcompose default in test/testdrive.

jkosh44 commented:
> > How did you run the Testdrive tests? The testdrive tests are somewhat sensitive to how they're run, so if it's not done properly there can be some false negatives.
> > The non-nightlies are passing so I think it's fine. The nightlies are still failing due to some issue with escaping characters.
> 
> Ah, I was running ./mzcompose default in test/testdrive.

I think sometimes if you don't run `./mzcompose down` then the next test won't start from a clean state.

jkosh44 reviewed (approved):
Thanks so much for the contribution! And thanks for slogging through all those test failures.
